sdks/java/io/hadoop-file-system/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystem.java

```diff
@@ -189,7 +189,21 @@ public int read(ByteBuffer dst) throws IOException {
       if (closed) {
         throw new IOException("Channel is closed");
       }
-      return inputStream.read(dst);
+      // O length read must be supported
+      int read = 0;
+      // We avoid using the ByteBuffer based read for Hadoop because some FSDataInputStream
+      // implementations are not ByteBufferReadable,
+      // See https://issues.apache.org/jira/browse/HADOOP-14603
+      if (dst.hasArray()){
+        // does the same as inputStream.read(dst):
+        // stores up to dst.remaining() bytes into dst.array() starting at dst.position().
+        // But dst can have an offset with its backing array hence the + dst.arrayOffset()
+        read = inputStream.read(dst.array(), dst.position() + dst.arrayOffset(), dst.remaining());
+      }
```

Comment:
Add an else block and either delegate to:
* inputStream.read(dst);
**OR**
* throw an UnsupportedOperationException since the ByteBuffer is off heap with a TODO saying add support for off heap byte buffers

This will prevent an infinite loop where the reader just continuously keeps getting back zero bytes.

Reply:
Yes of course, else is needed!
I prefer falling back to `inputStream.read(dst);` because its implementation is `((ByteBufferReadable)inputStream).read(byteBuffer);`  if the user were to pass an `InputStream` that is  `ByteBufferReadable` and which implementation of `read()` supports off-heap `byteBuffer`, then it will work.